File: backend/app/schemas/finding.py
"""Public Phase 8 finding requests, filters, and response schemas."""

from __future__ import annotations

import re
from datetime import date, datetime
from typing import Annotated, Any, Literal
from uuid import UUID

from pydantic import Field, field_validator

from app.models.compliance_enums import (
    FindingCode,
    FindingSeverity,
    FindingStatus,
    FindingType,
)
from app.schemas.base import ApiSchema
from app.schemas.common import PaginationData
from app.schemas.compliance import (
    ComplianceDocumentReference,
    ComplianceRevisionReference,
    ComplianceRuleReference,
)
from app.schemas.document_revision import UserReference

_CONTROL_CHARACTERS = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")


def sanitize_required_text(value: str) -> str:
    """Trim user text and reject hidden control characters."""
    normalized = value.strip()
    if not normalized:
        raise ValueError("Value must not be empty.")
    if _CONTROL_CHARACTERS.search(normalized):
        raise ValueError("Value contains unsupported control characters.")
    return normalized


def sanitize_optional_text(value: str | None) -> str | None:
    if value is None:
        return None
    normalized = value.strip()
    if not normalized:
        return None
    if _CONTROL_CHARACTERS.search(normalized):
        raise ValueError("Value contains unsupported control characters.")
    return normalized


class FindingCreateManualRequest(ApiSchema):
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    severity: FindingSeverity
    title: str = Field(min_length=1, max_length=500)
    description: str = Field(min_length=1, max_length=10_000)
    recommendation: str | None = Field(default=None, max_length=5000)
    source_reference: str | None = Field(default=None, max_length=1000)
    page_number: int | None = Field(default=None, ge=1)
    worksheet_name: str | None = Field(default=None, max_length=255)
    cell_coordinate: str | None = Field(default=None, max_length=50)
    language_code: str | None = Field(default=None, max_length=20)
    location: dict[str, Any] = Field(default_factory=dict)

    _title = field_validator("title", mode="before")(sanitize_required_text)
    _description = field_validator("description", mode="before")(sanitize_required_text)
    _optional_text = field_validator(
        "recommendation",
        "source_reference",
        "worksheet_name",
        "cell_coordinate",
        "language_code",
        mode="before",
    )(sanitize_optional_text)


class FindingUpdateRequest(ApiSchema):
    severity: FindingSeverity | None = None
    title: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = Field(
        default=None,
        min_length=1,
        max_length=10_000,
    )
    recommendation: str | None = Field(default=None, max_length=5000)

    @field_validator("title", "description", mode="before")
    @classmethod
    def required_when_present(cls, value: object) -> object:
        return sanitize_required_text(value) if isinstance(value, str) else value

    _recommendation = field_validator("recommendation", mode="before")(
        sanitize_optional_text
    )


class FindingReviewRequest(ApiSchema):
    comment: str = Field(min_length=1, max_length=5000)

    _comment = field_validator("comment", mode="before")(sanitize_required_text)


class FindingResolveRequest(ApiSchema):
    comment: str = Field(min_length=1, max_length=5000)

    _comment = field_validator("comment", mode="before")(sanitize_required_text)


class FindingReopenRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=5000)

    _reason = field_validator("reason", mode="before")(sanitize_required_text)


class FindingFalsePositiveRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=5000)

    _reason = field_validator("reason", mode="before")(sanitize_required_text)


class FindingAcceptRiskRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=5000)
    expiry_date: date

    _reason = field_validator("reason", mode="before")(sanitize_required_text)


class FindingAssignRequest(ApiSchema):
    assigned_to: UUID


class FindingBulkActionBase(ApiSchema):
    finding_ids: list[UUID] = Field(min_length=1, max_length=10_000)

    @field_validator("finding_ids")
    @classmethod
    def deduplicate_finding_ids(cls, value: list[UUID]) -> list[UUID]:
        """Preserve caller order while preventing duplicate mutations."""

        return list(dict.fromkeys(value))


class FindingBulkAssignRequest(FindingBulkActionBase):
    action: Literal["ASSIGN"]
    assigned_to: UUID


class FindingBulkReviewRequest(FindingBulkActionBase):
    action: Literal["REVIEW"]
    comment: str = Field(min_length=1, max_length=5000)

    _comment = field_validator("comment", mode="before")(sanitize_required_text)


FindingBulkActionRequest = Annotated[
    FindingBulkAssignRequest | FindingBulkReviewRequest,
    Field(discriminator="action"),
]


class FindingBulkActionResponse(ApiSchema):
    action: Literal["ASSIGN", "REVIEW"]
    processed_count: int = Field(ge=1)
    finding_ids: list[UUID] = Field(min_length=1)


class FindingOccurrenceResponse(ApiSchema):
    id: UUID
    finding_id: UUID
    compliance_run_id: UUID
    detected_at: datetime
    source_reference: str | None
    location: dict[str, Any] = Field(default_factory=dict)
    metrics: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class FindingHistoryEntry(ApiSchema):
    """One bounded workflow event derived from the append-only audit log."""

    id: UUID
    action: str
    previous_status: FindingStatus | None
    new_status: FindingStatus
    comment: str | None
    reason: str | None
    actor: UserReference | None
    created_at: datetime


class FindingListItem(ApiSchema):
    id: UUID
    compliance_run_id: UUID | None
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    finding_code: FindingCode
    finding_type: FindingType
    severity: FindingSeverity
    status: FindingStatus
    document: ComplianceDocumentReference | None = None
    revision: ComplianceRevisionReference | None = None
    validation_rule: ComplianceRuleReference | None = None
    title: str
    language_code: str | None
    detected_section_id: UUID | None
    section_code: str | None = None
    source_reference: str | None
    page_number: int | None
    worksheet_name: str | None
    cell_coordinate: str | None
    assigned_to: UserReference | UUID | None
    is_system_generated: bool
    is_repeat: bool
    created_at: datetime
    updated_at: datetime


class FindingResponse(FindingListItem):
    validation_rule_id: UUID | None
    description: str
    recommendation: str | None
    container_id: UUID | None
    translation_group_id: UUID | None
    extracted_block_id: UUID | None
    ocr_block_id: UUID | None
    location: dict[str, Any] = Field(default_factory=dict)
    expected_value: dict[str, Any] | list[Any] | None
    actual_value: dict[str, Any] | list[Any] | None
    metrics: dict[str, Any] = Field(default_factory=dict)
    previous_finding_id: UUID | None
    created_by: UUID | None
    reviewed_by: UserReference | UUID | None
    reviewed_at: datetime | None
    review_comment: str | None
    resolved_by: UserReference | UUID | None
    resolved_at: datetime | None
    resolution_comment: str | None
    false_positive_by: UserReference | UUID | None
    false_positive_at: datetime | None
    false_positive_reason: str | None
    accepted_risk_by: UserReference | UUID | None
    accepted_risk_at: datetime | None
    accepted_risk_reason: str | None
    accepted_risk_expiry_date: date | None
    reopened_by: UserReference | UUID | None
    reopened_at: datetime | None
    reopen_reason: str | None
    occurrences: list[FindingOccurrenceResponse] = Field(default_factory=list)
    history: list[FindingHistoryEntry] = Field(default_factory=list)


class FindingListResponse(PaginationData[FindingListItem]):
    pass


class FindingFilter(ApiSchema):
    search: str | None = None
    department_id: UUID | None = None
    document_id: UUID | None = None
    revision_id: UUID | None = None
    compliance_run_id: UUID | None = None
    detected_section_id: UUID | None = None
    section_code: str | None = None
    finding_code: FindingCode | None = None
    finding_type: FindingType | None = None
    severity: FindingSeverity | None = None
    status: FindingStatus | None = None
    language_code: str | None = None
    assigned_to: UUID | None = None
    created_by_system: bool | None = None
    created_from: datetime | None = None
    created_to: datetime | None = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    sort_by: str = "severity"
    sort_order: str = "desc"

File: frontend/src/types/finding.ts
import type {
  ComplianceDocumentSummary,
  ComplianceRevisionSummary,
  RequiredLanguageCode,
} from './compliance';
import type { DocumentUserSummary } from './document';
import type { PaginatedData, SortOrder } from './masterData';

export const findingSeverities = ['CRITICAL', 'MAJOR', 'MINOR', 'INFORMATION'] as const;

export type FindingSeverity = (typeof findingSeverities)[number];

export const findingStatuses = [
  'OPEN',
  'IN_REVIEW',
  'RESOLVED',
  'CLOSED',
  'FALSE_POSITIVE',
  'ACCEPTED_RISK',
  'REOPENED',
] as const;

export type FindingStatus = (typeof findingStatuses)[number];

export const findingTypes = [
  'DOCUMENT_CODE',
  'LANGUAGE_PRESENCE',
  'LANGUAGE_COVERAGE',
  'SECTION_MISSING',
  'SECTION_LANGUAGE_MISSING',
  'SECTION_ORDER',
  'LANGUAGE_ORDER',
  'TRANSLATION_GROUP_INCOMPLETE',
  'TABLE_LANGUAGE_MISSING',
  'CELL_LANGUAGE_MISSING',
  'UNKNOWN_LANGUAGE_EXCESS',
  'MIXED_LANGUAGE_EXCESS',
  'OCR_CONFIDENCE',
  'EXTRACTION_QUALITY',
  'STRUCTURE',
  'MANUAL',
] as const;

export type FindingType = (typeof findingTypes)[number];

export interface FindingReference {
  id: string;
  code: string;
  name: string;
}

export interface FindingHistoryEntry {
  id: string;
  action: string;
  previousStatus: FindingStatus | null;
  newStatus: FindingStatus;
  comment: string | null;
  reason: string | null;
  actor: DocumentUserSummary | null;
  createdAt: string;
}

export interface Finding {
  id: string;
  complianceRunId: string | null;
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  validationRuleId: string | null;
  document: ComplianceDocumentSummary | null;
  revision: ComplianceRevisionSummary | null;
  validationRule: FindingReference | null;
  findingCode: string;
  findingType: FindingType;
  severity: FindingSeverity;
  status: FindingStatus;
  title: string;
  description: string;
  recommendation: string | null;
  containerId: string | null;
  detectedSectionId: string | null;
  sectionCode: string | null;
  translationGroupId: string | null;
  extractedBlockId: string | null;
  ocrBlockId: string | null;
  pageNumber: number | null;
  worksheetName: string | null;
  cellCoordinate: string | null;
  sourceReference: string | null;
  location: Record<string, unknown> | null;
  languageCode: RequiredLanguageCode | null;
  expectedValue: unknown;
  actualValue: unknown;
  metrics: Record<string, unknown> | null;
  isSystemGenerated: boolean;
  isRepeat: boolean;
  previousFindingId: string | null;
  assignedTo: DocumentUserSummary | null;
  reviewedBy: DocumentUserSummary | null;
  reviewedAt: string | null;
  reviewComment: string | null;
  resolvedBy: DocumentUserSummary | null;
  resolvedAt: string | null;
  resolutionComment: string | null;
  falsePositiveBy: DocumentUserSummary | null;
  falsePositiveAt: string | null;
  falsePositiveReason: string | null;
  acceptedRiskBy?: DocumentUserSummary | null;
  acceptedRiskAt?: string | null;
  acceptedRiskReason?: string | null;
  acceptedRiskExpiryDate?: string | null;
  reopenedBy: DocumentUserSummary | null;
  reopenedAt: string | null;
  reopenReason: string | null;
  history: FindingHistoryEntry[];
  createdAt: string;
  updatedAt: string;
}

export type FindingListItem = Omit<
  Finding,
  | 'expectedValue'
  | 'actualValue'
  | 'metrics'
  | 'history'
  | 'reviewComment'
  | 'resolutionComment'
  | 'falsePositiveReason'
  | 'reopenReason'
>;

export interface FindingListParams {
  search?: string;
  departmentId?: string;
  documentId?: string;
  revisionId?: string;
  complianceRunId?: string;
  detectedSectionId?: string;
  findingCode?: string;
  findingType?: FindingType;
  severity?: FindingSeverity;
  status?: FindingStatus;
  languageCode?: RequiredLanguageCode;
  section?: string;
  assignedTo?: string;
  createdBySystem?: boolean;
  createdFrom?: string;
  createdTo?: string;
  page: number;
  pageSize: number;
  sortBy?: string;
  sortOrder?: SortOrder;
}

export type FindingList = PaginatedData<FindingListItem>;

export interface ManualFindingRequest {
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  severity: FindingSeverity;
  title: string;
  description: string;
  recommendation?: string | null;
  sourceReference?: string | null;
  pageNumber?: number | null;
  worksheetName?: string | null;
  cellCoordinate?: string | null;
}

export interface FindingUpdateRequest {
  severity?: FindingSeverity;
  title?: string;
  description?: string;
  recommendation?: string | null;
}

export interface FindingReviewRequest {
  comment: string;
}

export interface FindingResolveRequest {
  comment: string;
}

export interface FindingReturnToOpenRequest {
  comment: string;
}

export interface FindingFalsePositiveRequest {
  reason: string;
}

export interface FindingAcceptRiskRequest {
  reason: string;
  expiryDate: string;
}

export interface FindingReopenRequest {
  reason: string;
}

export interface FindingAssignRequest {
  assignedTo: string;
}

export type FindingBulkActionRequest =
  | {
      action: 'ASSIGN';
      findingIds: string[];
      assignedTo: string;
    }
  | {
      action: 'REVIEW';
      findingIds: string[];
      comment: string;
    };

export interface FindingBulkActionResult {
  action: FindingBulkActionRequest['action'];
  processedCount: number;
  findingIds: string[];
}

export interface FindingDownload {
  blob: Blob;
  fileName: string | null;
}

export type FindingExportFormat = 'json' | 'xlsx';

export interface FindingsReportSummary {
  totalFindings: number;
  open: number;
  inReview: number;
  resolved: number;
  critical: number;
  major: number;
  minor: number;
  information: number;
  falsePositive: number;
  acceptedRisk: number;
  byDepartment: Array<{ label: string; count: number }>;
  byType: Array<{ label: string; count: number }>;
  bySeverity: Array<{ label: FindingSeverity; count: number }>;
  trend: Array<{ period: string; count: number }>;
}

export interface FindingsReport {
  summary: FindingsReportSummary;
  findings: FindingList;
}

export const findingActionTransitions: Readonly<
  Record<
    FindingStatus,
    readonly (
      | 'review'
      | 'return-to-open'
      | 'resolve'
      | 'false-positive'
      | 'accept-risk'
      | 'reopen'
      | 'assign'
    )[]
  >
> = {
  OPEN: ['review', 'resolve', 'false-positive', 'accept-risk', 'assign'],
  IN_REVIEW: ['return-to-open', 'resolve', 'false-positive', 'accept-risk', 'assign'],
  RESOLVED: ['reopen', 'assign'],
  CLOSED: ['assign'],
  FALSE_POSITIVE: ['reopen', 'assign'],
  ACCEPTED_RISK: ['reopen', 'assign'],
  REOPENED: ['review', 'resolve', 'false-positive', 'accept-risk', 'assign'],
};
